### PR TITLE
Fixed snow being placed through slabs

### PR DESCRIPTION
**Pull request by me4502**
Uses the game's internal "canPlace" system to determine whether snow is capable of being placed at the location.

Adapters PR: https://github.com/EngineHub/worldedit-adapters/pull/21

Fixes https://github.com/EngineHub/WorldEdit/issues/1516

### CODE REVIEW

File: worldedit-core/src/main/java/com/sk89q/worldedit/function/block/SnowSimulator.java [hunk 1]
```diff
@@ -67,8 +68,7 @@ public boolean isGround(BlockVector3 position) {
         }
 
         // Can only place on full solid blocks
```
Comment (octylFractal) [inline worldedit-core/src/main/java/com/sk89q/worldedit/function/block/SnowSimulator.java:70]: Update the comment plox.

File: worldedit-core/src/main/java/com/sk89q/worldedit/function/block/SnowSimulator.java [hunk 2]
```diff
@@ -116,6 +116,9 @@ public boolean apply(BlockVector3 position, int depth) throws WorldEditException
             }
             return false;
         }
+        if (this.extent instanceof World && !((World) this.extent).canPlaceAt(abovePosition, snow)) {
```
Comment (octylFractal) [inline worldedit-core/src/main/java/com/sk89q/worldedit/function/block/SnowSimulator.java:119]: Just take and store `World`, if we want to use `Extent` in the future we can just change the constructor (that's backwards-compatible).

**Review by octylFractal — CHANGES_REQUESTED** [review]

File: worldedit-core/src/main/java/com/sk89q/worldedit/function/block/SnowSimulator.java (hunk 2 above)
Comment (me4502) [reply worldedit-core/src/main/java/com/sk89q/worldedit/function/block/SnowSimulator.java:119]: Due to the extent here being the one that applies blocks, we'd need to pass in a second world to do that. Instead, I've made this require a world or EditSession to be passed in

File: worldedit-core/src/main/java/com/sk89q/worldedit/function/block/SnowSimulator.java [hunk 3]
```diff
@@ -42,6 +46,10 @@
     private int affected;
 
     public SnowSimulator(Extent extent, boolean stack) {
+        checkArgument(extent instanceof World
```
Comment (octylFractal) [inline worldedit-core/src/main/java/com/sk89q/worldedit/function/block/SnowSimulator.java:49]: Unwrap the world into a separate field in constructor maybe?

**Review by octylFractal — CHANGES_REQUESTED** [review]

**Review by octylFractal — APPROVED** [review]

**Pull request merged**